daisy_workflows/build-publish/rhui/health_check.py

```diff
@@ -0,0 +1,131 @@
+#  Copyright 2022 Google Inc. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http:#www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Health check for RHUI infrastructure nodes."""
+
+import argparse
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+import typing
+
+
+class HealthChecks:
+  nfs_mount: str
+
+  def __init__(
+      self,
+      nfs_mount: str,
+  ) -> None:
+    self.nfs_mount = nfs_mount
+
+  def mount_uses_nfs(self, logger: logging.Logger):
+    logger.info("checking that %s is mounted as NFS", self.nfs_mount)
+    # Using pipes to capture output since RHEL8 has Python 3.6
+    # which doesn't support `capture_output`.
+    p = subprocess.run(
+      ["stat", "-fc", "%T", self.nfs_mount],
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE,
+    )
+    if p.returncode != 0:
+      raise EnvironmentError("Failed to stat %s: %s" % (self.nfs_mount, p))
+    mnt_type = p.stdout.decode("utf8").strip()
+    if mnt_type != "nfs":
+      raise EnvironmentError("Not mounted as nfs. stat results: %s" % p)
+
+  def mount_is_writable(self, logger: logging.Logger):
+    logger.info("checking that %s is writable", self.nfs_mount)
+    with tempfile.TemporaryFile(dir=self.nfs_mount) as f:
+      written_bytes = b"content"
+      f.write(written_bytes)
+      f.seek(0)
+      actual = f.read()
+      if actual != written_bytes:
+        logger.info(
+          "expected file to contain %s actual %s", written_bytes, actual
+        )
+        raise EnvironmentError("Failed to write to %s" % self.nfs_mount)
+
+  def mount_is_readable(self, logger: logging.Logger):
+    logger.info("checking that %s is readable", self.nfs_mount)
+    os.listdir(self.nfs_mount)
+
+
+def main(node_type: str, result_file: typing.TextIO, nfs_mount: str):
+  """Executes health check for node_type, and writes the overall
+  result to result_file ('OK' if all checks pass, otherwise 'ERROR').
+  """
+  health_checks = HealthChecks(nfs_mount=nfs_mount)
+  checks = [
+    health_checks.mount_is_readable,
+    health_checks.mount_uses_nfs,
+  ]
+  if node_type == "rhua":
+    checks += [
+      health_checks.mount_is_writable,
+    ]
+  success = True
+  for func in checks:
+    logger = logging.getLogger(func.__name__)
```

Comment:
is the idea that we would get named output logs without having the health checks themselves do logging?

Reply:
Yes -- each health check gets a logger that's scoped to its name. Also, this loop handles logging success and failure messages.

Here's some examples from journalctl:

## Prior to mounting NFS:

The error messages appear in read, and they are filterable by with `journalctl -p 3`

```
Mar 24 18:30:46 rhel-8 systemd[1]: Starting RHUA health check...
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_is_readable - INFO - checking that /var/lib/rhui/remote_share is readable
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_is_readable - ERROR - [Errno 2] No such file or directory: '/var/lib/rhui/remote_share'
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_uses_nfs - INFO - checking that /var/lib/rhui/remote_share is mounted as NFS
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_uses_nfs - ERROR - /var/lib/rhui/remote_share not in /proc/mounts
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_is_writable - INFO - checking that /var/lib/rhui/remote_share is writable
Mar 24 18:30:46 rhel-8 rhui-health-check[1916]: mount_is_writable - ERROR - [Errno 2] No such file or directory: '/var/lib/rhui/remote_share/tmp20xplcm1'
Mar 24 18:30:46 rhel-8 systemd[1]: rhui-health-check.service: Succeeded.
```

## After mounting NFS:

```
Mar 24 18:33:14 rhel-8 systemd[1]: Starting RHUA health check...
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_is_readable - INFO - checking that /var/lib/rhui/remote_share is readable
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_is_readable - INFO - success
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_uses_nfs - INFO - checking that /var/lib/rhui/remote_share is mounted as NFS
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_uses_nfs - INFO - success
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_is_writable - INFO - checking that /var/lib/rhui/remote_share is writable
Mar 24 18:33:14 rhel-8 rhui-health-check[1966]: mount_is_writable - INFO - success
Mar 24 18:33:14 rhel-8 systemd[1]: rhui-health-check.service: Succeeded.
```